platforms/core-runtime/launcher/src/integTest/groovy/org/gradle/launcher/CommandLineOutputIntegrationTest.groovy

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.internal.buildconfiguration.DaemonJvmPropertiesDefaults
+import org.gradle.internal.jvm.Jvm
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.gradle.test.precondition.Requires
+import org.gradle.test.preconditions.IntegTestPreconditions
+import org.junit.Rule
+
+/**
+ * Assertions over the output of certain command line invocations.
+ */
+class CommandLineOutputIntegrationTest extends AbstractIntegrationSpec {
```

Comment:
❌ I was thinking something a little simpler.  I don't think we care about the effects of --project-dir or running from another working directory because we're using the same mechanisms that read `gradle.properties` or find the build files.  

If you think these _are_ important, I would slice the tests differently.  When the build's JVM is based on org.gradle.java.home or the launcher, we know the directories can't affect things.  If we want to see what happens with `--project-dir`, we can just test it with the JVM criteria.

Two other more generic things:
1. We should try to avoid disabling tests for the embedded executor unless it really doesn't make sense to run them. This makes it easier for people to just run with the embedded executor most of the time and the tests that need to run outside of it can.  In my example below, that's done by enabling daemons (we don't start daemons for this test, but that's enough to disable the in-process executor).
2. We need to be careful to restrict toolchain discovery tests because they can look everywhere. We have some fixtures that let you limit which JDKs are visible to the tests. My example does this.  It's also good to mark these as requirements for the test and write it in a way that doesn't require a specific JVM version. 

```
class CommandLineOutputIntegrationTest extends AbstractIntegrationSpec implements JavaToolchainFixture, DaemonJvmPropertiesFixture {

    def setup() {
        // forces the tests to fork and not run in-process
        executer.requireDaemon().requireIsolatedDaemons()
    }

    def "displays version message appropriately for Build JVM requested by Launcher JVM"() {
        when:
        succeeds("--version")

        then:
        outputContains("Requested Daemon JVM:  ${Jvm.current().javaHome.absolutePath} (no JDK specified, using Launcher JVM)")
    }

    def "displays version message appropriately for Build JVM requested by org.gradle.java.home"() {
        given:
        file("gradle.properties").writeProperties("org.gradle.java.home": Jvm.current().javaHome.absolutePath)

        when:
        succeeds("--version")

        then:
        outputContains("Requested Daemon JVM:  '${Jvm.current().javaHome.absolutePath}' (from -Dorg.gradle.java.home)")
    }

    @Requires(IntegTestPreconditions.JavaHomeWithDifferentVersionAvailable)
    def "displays version message appropriately for Build JVM requested by build criteria"() {
        def otherJvm = AvailableJavaHomes.differentVersion

        given:
        writeJvmCriteria(otherJvm)

        when:
        withInstallations(otherJvm).succeeds("--version")

        then:
        outputContains("Requested Daemon JVM:  Compatible with Java ${otherJvm.javaVersion.majorVersion} (Daemon JVM criteria from gradle/gradle-daemon-jvm.properties)")
    }
}
```